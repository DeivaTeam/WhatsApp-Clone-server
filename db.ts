import 'reflect-metadata';
import { Chat } from './modules/recipient/models/Chat';
import { Recipient } from './modules/recipient/models/Recipient';
import moment from 'moment';
import { Message } from './modules/recipient/models/Message';
import { User } from './modules/recipient/models/User';
import { Connection } from 'typeorm';
import AccountsPassword from '@accounts/password';
import { hashPassword } from '@accounts/password/lib/utils';

export enum MessageType {
  PICTURE,
  TEXT,
  LOCATION,
}

export async function addSampleData(connection: Connection, accountsPassword: AccountsPassword) {

  await connection.synchronize();

  const userId1 = await accountsPassword.createUser({
    username: 'ethan',
    password: hashPassword('111', 'sha256'),
    name: 'Ethan Gonzalez',
    picture: 'https://randomuser.me/api/portraits/thumb/men/1.jpg',
    phone: '+391234567890',
  });

  const userId2 = await accountsPassword.createUser({
    username: 'bryan',
    password: hashPassword('222', 'sha256'),
    name: 'Bryan Wallace',
    picture: 'https://randomuser.me/api/portraits/thumb/men/2.jpg',
    phone: '+391234567891',
  });

  const userId3 = await accountsPassword.createUser({
    username: 'avery',
    password: hashPassword('333', 'sha256'),
    name: 'Avery Stewart',
    picture: 'https://randomuser.me/api/portraits/thumb/women/1.jpg',
    phone: '+391234567892',
  });

  const userId4 = await accountsPassword.createUser({
    username: 'katie',
    password: hashPassword('444', 'sha256'),
    name: 'Katie Peterson',
    picture: 'https://randomuser.me/api/portraits/thumb/women/2.jpg',
    phone: '+391234567893',
  });

  const userId5 = await accountsPassword.createUser({
    username: 'ray',
    password: hashPassword('555', 'sha256'),
    name: 'Ray Edwards',
    picture: 'https://randomuser.me/api/portraits/thumb/men/3.jpg',
    phone: '+391234567894',
  });

  const userId6 = await accountsPassword.createUser({
    username: 'niko',
    password: hashPassword('666', 'sha256'),
    name: 'Niccol√≤ Belli',
    picture: 'https://randomuser.me/api/portraits/thumb/men/4.jpg',
    phone: '+391234567895',
  });

  const userId7 = await accountsPassword.createUser({
    username: 'mario',
    password: hashPassword('777', 'sha256'),
    name: 'Mario Rossi',
    picture: 'https://randomuser.me/api/portraits/thumb/men/5.jpg',
    phone: '+391234567896',
  });
  
  const user1 = await connection.getRepository(User).findOne(userId1) as User;
  const user2 = await connection.getRepository(User).findOne(userId2) as User;
  const user3 = await connection.getRepository(User).findOne(userId3) as User;
  const user4 = await connection.getRepository(User).findOne(userId4) as User;
  const user5 = await connection.getRepository(User).findOne(userId5) as User;
  const user6 = await connection.getRepository(User).findOne(userId6) as User;
  const user7 = await connection.getRepository(User).findOne(userId7) as User;


  await connection.manager.save(new Chat({
    allTimeMembers: [user1, user3],
    listingMembers: [user1, user3],
    messages: [
      new Message({
        sender: user1,
        content: 'You on your way?',
        createdAt: moment().subtract(1, 'hours').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user3],
        recipients: [
          new Recipient({
            user: user3,
          }),
        ],
      }),
      new Message({
        sender: user3,
        content: 'Yep!',
        createdAt: moment().subtract(1, 'hours').add(5, 'minutes').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user3],
        recipients: [
          new Recipient({
            user: user1,
          }),
        ],
      }),
    ],
  }));

  await connection.manager.save(new Chat({
    allTimeMembers: [user1, user4],
    listingMembers: [user1, user4],
    messages: [
      new Message({
        sender: user1,
        content: 'Hey, it\'s me',
        createdAt: moment().subtract(2, 'hours').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user4],
        recipients: [
          new Recipient({
            user: user4,
          }),
        ],
      }),
    ],
  }));

  await connection.manager.save(new Chat({
    allTimeMembers: [user1, user5],
    listingMembers: [user1, user5],
    messages: [
      new Message({
        sender: user1,
        content: 'I should buy a boat',
        createdAt: moment().subtract(1, 'days').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user5],
        recipients: [
          new Recipient({
            user: user5,
          }),
        ],
      }),
      new Message({
        sender: user1,
        content: 'You still there?',
        createdAt: moment().subtract(1, 'days').add(16, 'hours').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user5],
        recipients: [
          new Recipient({
            user: user5,
          }),
        ],
      }),
    ],
  }));

  await connection.manager.save(new Chat({
    allTimeMembers: [user3, user4],
    listingMembers: [user3, user4],
    messages: [
      new Message({
        sender: user3,
        content: 'Look at my mukluks!',
        createdAt: moment().subtract(4, 'days').toDate(),
        type: MessageType.TEXT,
        holders: [user3, user4],
        recipients: [
          new Recipient({
            user: user4,
          }),
        ],
      }),
    ],
  }));

  await connection.manager.save(new Chat({
    allTimeMembers: [user2, user5],
    listingMembers: [user2, user5],
    messages: [
      new Message({
        sender: user2,
        content: 'This is wicked good ice cream.',
        createdAt: moment().subtract(2, 'weeks').toDate(),
        type: MessageType.TEXT,
        holders: [user2, user5],
        recipients: [
          new Recipient({
            user: user5,
          }),
        ],
      }),
      new Message({
        sender: user5,
        content: 'Love it!',
        createdAt: moment().subtract(2, 'weeks').add(10, 'minutes').toDate(),
        type: MessageType.TEXT,
        holders: [user2, user5],
        recipients: [
          new Recipient({
            user: user2,
          }),
        ],
      }),
    ],
  }));

  await connection.manager.save(new Chat({
    allTimeMembers: [user1, user6],
    listingMembers: [user1],
  }));

  await connection.manager.save(new Chat({
    allTimeMembers: [user2, user1],
    listingMembers: [user2],
  }));

  await connection.manager.save(new Chat({
    name: 'Ethan\'s group',
    picture: 'https://randomuser.me/api/portraits/thumb/lego/1.jpg',
    allTimeMembers: [user1, user3, user4, user6],
    listingMembers: [user1, user3, user4, user6],
    actualGroupMembers: [user1, user4, user6],
    admins: [user1, user6],
    owner: user1,
    messages: [
      new Message({
        sender: user1,
        content: 'I made a group',
        createdAt: moment().subtract(2, 'weeks').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user3, user4, user6],
        recipients: [
          new Recipient({
            user: user3,
          }),
          new Recipient({
            user: user4,
          }),
          new Recipient({
            user: user6,
          }),
        ],
      }),
      new Message({
        sender: user1,
        content: 'Ops, Avery was not supposed to be here',
        createdAt: moment().subtract(2, 'weeks').add(2, 'minutes').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user4, user6],
        recipients: [
          new Recipient({
            user: user4,
          }),
          new Recipient({
            user: user6,
          }),
        ],
      }),
      new Message({
        sender: user4,
        content: 'Awesome!',
        createdAt: moment().subtract(2, 'weeks').add(10, 'minutes').toDate(),
        type: MessageType.TEXT,
        holders: [user1, user4, user6],
        recipients: [
          new Recipient({
            user: user1,
          }),
          new Recipient({
            user: user6,
          }),
        ],
      }),
    ],
  }));

  await connection.manager.save(new Chat({
    name: 'Ray\'s group',
    allTimeMembers: [user3, user6],
    listingMembers: [user3, user6],
    actualGroupMembers: [user3, user6],
    admins: [user6],
    owner: user6,
  }));
}
