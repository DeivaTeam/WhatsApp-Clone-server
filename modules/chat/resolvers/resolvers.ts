import { PubSub, withFilter } from 'apollo-server-express';
import { Chat } from '../models/Chat';
import { ChatProvider } from '../providers/chat.provider';
import { ModuleContext } from '@graphql-modules/core';

export const Subscription = {
  chatAdded: {
    subscribe: withFilter((root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('chatAdded'),
      (data: { chatAdded: Chat, creatorId: number }, variables, { injector }: ModuleContext) =>
        data && injector.get(ChatProvider).filterChatAddedOrUpdated(data.chatAdded, data.creatorId)
    ),
  },
  chatUpdated: {
    subscribe: withFilter((root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('chatUpdated'),
      (data: { chatUpdated: Chat, updaterId: number }, variables, { injector }: ModuleContext) =>
        data && injector.get(ChatProvider).filterChatAddedOrUpdated(data.chatUpdated, data.updaterId)
    ),
  },
}

