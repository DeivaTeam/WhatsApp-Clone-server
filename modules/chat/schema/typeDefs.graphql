type Query {
  chats: [Chat!]! @auth
  chat(chatId: ID!): Chat @auth
}

type Subscription {
  chatAdded: Chat
  chatUpdated: Chat
}

type Chat {
  #May be a chat or a group
  id: ID!
  createdAt: Date!
  #Computed for chats
  name: String
  #Computed for chats
  picture: String
  #All members, current and past ones. Includes users who still didn't get the chat listed.
  allTimeMembers: [User!]!
  #Whoever gets the chat listed. For groups includes past members who still didn't delete the group. For chats they are the only ones who can send messages.
  listingMembers: [User!]!
  #Actual members of the group. Null for chats. For groups they are the only ones who can send messages. They aren't the only ones who get the group listed.
  actualGroupMembers: [User!]
  #Null for chats
  admins: [User!]
  #If null the group is read-only. Null for chats.
  owner: User
  #Computed property
  isGroup: Boolean!
}

type Mutation {
  addChat(userId: ID!): Chat @auth
  addGroup(userIds: [ID!]!, groupName: String!, groupPicture: String): Chat @auth
  updateGroup(chatId: ID!, groupName: String, groupPicture: String): Chat @auth
  removeChat(chatId: ID!): ID @auth
  addAdmins(groupId: ID!, userIds: [ID!]!): [ID]! @auth
  removeAdmins(groupId: ID!, userIds: [ID!]!): [ID]! @auth
  addMembers(groupId: ID!, userIds: [ID!]!): [ID]! @auth
  removeMembers(groupId: ID!, userIds: [ID!]!): [ID]! @auth
}
