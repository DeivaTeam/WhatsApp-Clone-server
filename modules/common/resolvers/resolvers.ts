import { GraphQLDateTime } from 'graphql-iso-date';
import { ScalarType } from 'graphql-toolkit';

export const Date = GraphQLDateTime;
