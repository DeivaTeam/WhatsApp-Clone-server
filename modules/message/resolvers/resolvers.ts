import { PubSub } from 'apollo-server-express';
import { withFilter } from 'apollo-server-express';
import { Message } from '../models/Message';
import { MessageProvider } from '../providers/message.provider';
import { ModuleContext } from '@graphql-modules/core';

export default {
  Subscription: {
    messageAdded: {
      subscribe: withFilter((root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('messageAdded'),
        (data: { messageAdded: Message }, variables, { injector }: ModuleContext) => data && injector.get(MessageProvider).filterMessageAdded(data.messageAdded)
      ),
    },
  },
};
