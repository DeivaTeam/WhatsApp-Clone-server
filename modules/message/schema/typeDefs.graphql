type Subscription {
  messageAdded: Message
}

enum MessageType {
  LOCATION
  TEXT
  PICTURE
}
