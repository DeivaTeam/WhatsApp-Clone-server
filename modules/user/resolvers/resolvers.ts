import { PubSub, GraphQLUpload } from 'apollo-server-express';
import { withFilter } from 'apollo-server-express';
import { UserProvider } from '../providers/user.provider';
import { User } from '../models/User';
import { ModuleContext } from '@graphql-modules/core';

export const Upload = GraphQLUpload;
export const Subscription = {
  userAdded: {
    subscribe: withFilter(
      (root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('userAdded'),
      (data: { userAdded: User }, variables, { injector }: ModuleContext) => data && injector.get(UserProvider).filterUserAddedOrUpdated(data.userAdded),
    ),
  },
  userUpdated: {
    subscribe: withFilter(
      (root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('userAdded'),
      (data: { userUpdated: User }, variables, { injector }: ModuleContext) => data && injector.get(UserProvider).filterUserAddedOrUpdated(data.userUpdated)
    ),
  },
}
