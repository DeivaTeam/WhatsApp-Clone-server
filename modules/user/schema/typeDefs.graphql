scalar Upload

type Subscription {
  userAdded: User
  userUpdated: User
}
