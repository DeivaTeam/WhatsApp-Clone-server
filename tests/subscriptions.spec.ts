import { ApolloServer, gql } from "apollo-server";
import typeDefs from "../schema/typeDefs";
import { resolvers } from "../schema/resolvers";
import ApolloClient from "apollo-client";
import { WebSocketLink } from "apollo-link-ws";
import { InMemoryCache, NormalizedCacheObject } from "apollo-cache-inmemory";
import ws from 'ws';
import { db } from "../db";

const GRAPHQL_ENDPOINT = 'ws://localhost:4000/graphql';

describe('Apollo Server', () => {
  let server: ApolloServer;
  let apolloClientUser1: ApolloClient<NormalizedCacheObject>;
  let apolloClientUser2: ApolloClient<NormalizedCacheObject>;

  beforeAll(async() => {
    server = new ApolloServer({
      typeDefs: gql(typeDefs),
      resolvers: resolvers as any,
      context: async ({ req, connection }: any) => connection ? connection.context : ({ currentUser: db.users[0] }),
      subscriptions: {
        onConnect: ({ currentUser }: any) => ({ currentUser }),
      },
    });

    await server.listen();

    apolloClientUser1 = new ApolloClient({
      link: new WebSocketLink({
        uri: GRAPHQL_ENDPOINT,
        webSocketImpl: ws,
        options: {
          connectionParams: () => ({ currentUser: db.users[0] }),
        },
      }),
      cache: new InMemoryCache(),
    });

    apolloClientUser2 = new ApolloClient({
      link: new WebSocketLink({
        uri: GRAPHQL_ENDPOINT,
        webSocketImpl: ws,
        options: {
          connectionParams: () => ({ currentUser: db.users[1] }),
        },
      }),
      cache: new InMemoryCache(),
    });
  });

  it('should match the snapshot for the GetChats query', async () => {
    const res: any = await new Promise(async (resolve, reject) => {
      apolloClientUser1.subscribe({
        query: gql`
            subscription messageAdded {
                messageAdded {
                    id
                    content
                    chat {
                        id,
                    },
                }
            }
        `,
      }).subscribe({
        next: resolve,
        error: reject,
      });

      await apolloClientUser2.mutate({
        mutation: gql`
            mutation AddMessage($chatId: ID!, $content: String!) {
                addMessage(chatId: $chatId, content: $content) {
                    id
                    content
                }
            }
        `,
        variables: { chatId: 7, content: 'This is a test' },
      });
    });

    expect(res.data).toBeDefined();
    expect(res.data.messageAdded).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
});
