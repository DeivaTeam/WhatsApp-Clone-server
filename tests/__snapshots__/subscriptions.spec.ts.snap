// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apollo Server should match the snapshot for the GetChats query 1`] = `
Object {
  "messageAdded": Object {
    "__typename": "Message",
    "chat": Object {
      "__typename": "Chat",
      "id": "7",
    },
    "content": "This is a test",
    "id": "1",
  },
}
`;
